grudge/dt_utils.py

```diff
@@ -45,22 +57,59 @@
 from pytools import memoize_on_first_arg
 
 
+def characteristic_lengthscales(dcoll: DiscretizationCollection) -> DOFArray:
+    r"""Computes the characteristic lengthscales at each node of a physical element.
+    The estimate is obtained using the following formula:
+
+    .. math::
+
+        h_{loc} = \operatorname{min}\left(\Delta r_i\right) r_D
+
+    where :math:`\operatorname{min}\left(\Delta r_i\right)` is the minimum
+    node distance on the reference cell (see :func:`dt_non_geometric_factors`),
+    and :math:`r_D` is the inradius of the cell (see :func:`dt_geometric_factors`).
+
+    :returns: a :class:`~meshmode.dof_array.DOFArray` containing a characteristic
+        lengthscale for each element, at each nodal location.
+
+    .. note::
+
+        While a prediction of stability is only meaningful in relation to a given
+        time integrator with a known stability region, the lengthscale provided here
+        is not intended to be specific to any one time integrator, though the
+        stability region of standard four-stage, fourth-order Runge-Kutta
+        methods has been used as a guide. Any concrete time integrator will
+        likely require scaling of the values returned by this routine.
+    """
+    return DOFArray(
+        dcoll._setup_actx,
```

Comment:
The setup array context should never escape into user expressions, unless you're keen on very difficult to debug synchronization bugs.

Some general rules:
* if the user did not explicitly supply an array context, a routine should only ever return frozen data.
* Thawed data (i.e. data that's attached to an array context) should never be memoized.

Clearly, this concept needs a better explanation. I'll work on that.

x-ref: https://github.com/inducer/arraycontext/issues/19

Reply:
https://github.com/inducer/arraycontext/pull/22